pytorch_lightning/trainer/trainer.py

```diff
@@ -943,24 +955,18 @@ def tune(
 
         return result
 
-    def _restore_modules_and_callbacks(self) -> None:
+    def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None) -> None:
         # restore modules after setup
-        if self.state.fn == TrainerFn.FITTING:
-            self.checkpoint_connector.resume_start()
-        self.checkpoint_connector.restore_datamodule()
+        self.checkpoint_connector.resume_start(checkpoint_path)
         self.checkpoint_connector.restore_model()
-        # restore callback states
-        self.checkpoint_connector.restore_callbacks()
-
-    def _load_checkpoint_weights(self):
-        # only one process running at this point for TPUs, as spawn isn't triggered yet
-        # todo: move this logic internally within the barrier.
-        if not self._device_type == DeviceType.TPU:
-            self.accelerator.barrier()
-        rank_zero_info(f"Loading model weights from checkpoint at {self._ckpt_path}")
-        self.checkpoint_connector.restore_model_weights(self._ckpt_path)
-
-    def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
+        if self.state.fn == TrainerFn.FITTING:
+            self.checkpoint_connector.restore_datamodule()
+            # restore callback states
+            self.checkpoint_connector.restore_callbacks()
```

Comment:
we should restore datamodule & callback states even if we're not fitting right?

Reply:
i was wondering this too. currently i think only fitting restores `restore_datamodule`, `restore_callbacks`, and `restore_training_state` as these fns all return early if `checkpoint_connector._loaded_checkpoint` is not set, `checkpoint_connector._loaded_checkpoint` is only set in `resume_start` which is only called for fitting?

Reply:
@awaelchli @carmocca @rohitgr7 do you know if this was intentional?

if not fitting is it expected to only load the weights, and not data/callback states? @rohitgr7 is fixing the trainer state when not fitting here: https://github.com/PyTorchLightning/pytorch-lightning/issues/9404

Reply:
in #9413 we are now loading the checkpoint to retrieve the `global_step` and `current_epoch` if not fitting and nothing else. Not sure about datamodule but for callbacks we never loaded any states if not fitting. we have an old issue to discuss whether to load the callback states or not here: https://github.com/PyTorchLightning/pytorch-lightning/issues/5542